### PR TITLE
Allow decoding Events containing BitVecs

### LINKED ISSUE #405: Handle BitVec decoding in Events
Broken out from https://github.com/paritytech/subxt/issues/309.

Currently if we encounter a `BitVec` when attempting to decode an event we crash:

https://github.com/paritytech/subxt/commit/0d07b0c91ec06a46e5ff5776f90d4cd7bda0b96d

We should support decoding `BitVec`, and/or more gracefully handle our inability to decode it as a workaround.

### PR DESCRIPTION
**Pull request by jsdw**
Also a little refactoring to make it easier to test the core decode logic, and remove the need to re-encode.

closes #405

### CODE REVIEW

**Comment by jsdw** [comment]
I had a look into the encode impl and reworked the handling to cover more cases and not care about ordering

File: src/events.rs [hunk 1]
```diff
@@ -161,165 +165,189 @@ impl<T: Config> EventsDecoder<T> {
...
-            output: &mut Vec<u8>,
-        ) -> Result<(), BasicError> {
-            let decoded = T::decode(input)?;
-            decoded.encode_to(output);
+        let all_bytes = *input;
+        // consume some bytes, moving the cursor forward:
+        decode_and_consume_type(type_id, &self.metadata.runtime_metadata().types, input)?;
+        // count how many bytes were consumed based on remaining length:
+        let consumed_len = all_bytes.len() - input.len();
+        // move those consumed bytes to the output vec unaltered:
```
Comment (ascjones) [inline src/events.rs:173]: Nice, much more efficient :muscle:

**Review by ascjones — APPROVED** [review]
Awesome!

File: src/events.rs [hunk 2]
```diff
@@ -161,165 +165,189 @@ impl<T: Config> EventsDecoder<T> {
...
+        output.extend(&all_bytes[0..consumed_len]);
+        Ok(())
+    }
+}
+
+// Given a type Id and a type registry, attempt to consume the bytes
+// corresponding to that type from our input.
+fn decode_and_consume_type(
```
Comment (jsdw) [inline src/events.rs:181]: TODO: If we are happy and merge this, we could create a new "good first issue" to review the tests, perhaps adding some like the bitvec ones to test this method more specifically?

Comment (ascjones) [reply src/events.rs:181]: Yes this would be a good method to unit test

File: src/events.rs [hunk 3]
```diff
@@ -404,6 +449,19 @@ mod tests {
         EventsDecoder::<DefaultConfig>::new(metadata)
     }
 
+    fn decode_and_consume_type_consumes_all_bytes<
```
Comment (TarikGul) [inline src/events.rs:452]: 👍

**Review by TarikGul — APPROVED** [review]

**Pull request merged**